### PR TITLE
Verified roblobs

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR sets some pieces on the board for a move towards consolidated blob verification and block signature caching.
- Adds `VerifiedROBlob` type to represent blobs that have undergone full verification (signed block header, inclusion proof valid, commitments verified). This allows us to leverage the type system to specify which paths deal with unverified vs verified values.
- Add `verification` package stub with eg `verification.BlobSidecarNoop` method to fake verification. This punts the overhaul of verification to a subsequent PR so we can make progress with the type system changes first and reduce the scope of the later verification PR.
- Refactor `BlobStorage`:
  -  Add a `Get`, and `Indices` helper method.
  - Simplify on-disk storage of blob data to reduce the number of system calls needed for the most common storage operations. 
  - Wire up storage in the node init with a flag to specify the path (default to a subdirectory in bolt data dir).
  - Use afero for simplified basepath handling and in-memory test db feature.
- Replace usages of boltdb blob methods with `BlobStorage` methods. Note that these methods only deal with `VerifiedROBlobs`, guaranteeing that anything we write/read from storage is verified.
- Replace numerous instances of `DepreatedBlobSidecar` with `ROBlob` or `VerifiedROBlob`.
- Fixes to unit tests to deal with these changes.
- Simplify `BlobAlignsWithBlock` because it had a ton of unit tests that weren't relevant with block fields moved into the header, which were also tricky to port, so it was just easier to fast-forward to replacing most of those with a block root check.

### CODE REVIEW

File: beacon-chain/node/node.go [hunk 1]
```diff
@@ -200,6 +203,14 @@ func New(cliCtx *cli.Context, opts ...Option) (*BeaconNode, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	log.Debugln("Starting Blob Storage")
+	blobStorage, err := filesystem.NewBlobStorage(beacon.BlobStoragePath)
```
Comment (kasey) [inline beacon-chain/node/node.go:208]: The initialization order of setting up the top-level context and calculating functional args makes it complicated to directly initialize the BlobStorage value as a functional arg. `BlobStoragePath` is a work-around for that complication, allowing the functional arg to set the path as a string, deferring actual initialization until the New method runs.

File: beacon-chain/sync/verify/blob.go [hunk 2]
```diff
@@ -2,70 +2,42 @@ package verify
...
-		return err
-	}
-
-	if len(commits) == 0 {
-		return nil
-	}
-
-	if blob.Index >= field_params.MaxBlobsPerBlock {
-		return errors.Wrapf(ErrIncorrectBlobIndex, "blob index %d >= max blobs per block %d", blob.Index, field_params.MaxBlobsPerBlock)
+	if blob.Index > fieldparams.MaxBlobsPerBlock {
```
Comment (prestonvanloon) [inline beacon-chain/sync/verify/blob.go:23]: Does index start at 0? If so, it should be `>=`

File: beacon-chain/db/filesystem/ephemeral.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+func NewEphemeralBlobStorage() *BlobStorage {
+	return &BlobStorage{fs: afero.NewMemMapFs()}
+}
+
+// NewEphemeralBlobStorageWithFs can be used by tests that want access to the virtual filesystem
+// in order to interact with it outside the parameters of the BlobStorage api.
+func NewEphemeralBlobStorageWithFs() (afero.Fs, *BlobStorage) {
+	fs := afero.NewMemMapFs()
+	return fs, &BlobStorage{fs: fs}
+}
```
Comment (terencechain) [inline beacon-chain/db/filesystem/ephemeral.go:17]: Is there a tag we can add to enforce this is test only and fail if it's used otherwise of test scope?

File: beacon-chain/sync/blobs_test.go [hunk 4]
```diff
@@ -103,32 +107,45 @@ func generateTestBlockWithSidecars(t *testing.T, parent [32]byte, slot types.Slo
...
 	for i, c := range block.Block.Body.BlobKzgCommitments {
-		sidecars[i] = generateTestSidecar(root, block, i, c)
+		sidecars[i] = generateTestSidecar(t, root, sbb, i, c)
 	}
 	return block, sidecars
 }
 
-func generateTestSidecar(root [32]byte, block *ethpb.SignedBeaconBlockDeneb, index int, commitment []byte) *ethpb.DeprecatedBlobSidecar {
+func generateTestSidecar(t *testing.T, root [32]byte, block interfaces.ReadOnlySignedBeaconBlock, index int, commitment []byte) blocks.ROBlob {
+	header, err := interfaces.SignedBeaconBlockHeaderFromBlockInterface(block)
```
Comment (terencechain) [inline beacon-chain/sync/blobs_test.go:120]: I think this is cleaner

```suggestion
	header, err := block.Header()
```

File: beacon-chain/sync/initial-sync/round_robin.go [hunk 5]
```diff
@@ -163,8 +164,14 @@ func (s *Service) processFetchedDataRegSync(
 	blksWithoutParentCount := 0
 	for _, b := range data.bwb {
 		if len(b.Blobs) > 0 {
-			if err := s.cfg.DB.SaveBlobSidecar(ctx, b.Blobs); err != nil {
-				log.WithError(err).Warn("Failed to save blob sidecar")
+			verified, err := verification.BlobSidecarSliceNoop(b.Blobs)
+			if err != nil {
+				log.WithField("root", b.Block.Root()).WithError(err).Error("blobs failed verification")
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/round_robin.go:169]: If the blob fails verification, why are we still saving it below?

File: beacon-chain/sync/validate_blob.go [hunk 6]
```diff
@@ -55,12 +53,15 @@ func (s *Service) validateBlob(ctx context.Context, pid peer.ID, msg *pubsub.Mes
...
 	if !ok {
-		log.WithField("message", m).Error("Message is not of type *eth.SignedBlobSidecar")
+		log.WithField("message", m).Error("Message is not of type *eth.BlobSidecar")
 		return pubsub.ValidationReject, errWrongMessage
 	}
-	blob := sBlob.Message
+	blob, err := blocks.NewROBlob(bpb)
+	if err != nil {
+		return pubsub.ValidationReject, errors.Wrap(err, "roblob conversion failure")
+	}
```
Comment (terencechain) [inline beacon-chain/sync/validate_blob.go:64]: I have a PR for these change but we should merge this first and I can deal with conflicts in my PR

File: beacon-chain/sync/verify/blob.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/sync/verify/blob.go:23]: Yep. It should be `>=`

File: beacon-chain/sync/verify/blob.go [hunk 7]
```diff
@@ -2,70 +2,42 @@ package verify
...
+	ErrIncorrectBlobIndex        = errors.New("incorrect blob index")
+	ErrBlobBlockMisaligned       = errors.Wrap(errBlobVerification, "root of block header in blob sidecar does not match block root")
 	ErrMismatchedBlobCommitments = errors.Wrap(errBlobVerification, "commitments at given slot, root and index do not match")
-	ErrMismatchedProposerIndex   = errors.Wrap(errBlobVerification, "proposer index does not match")
-	ErrIncorrectBlobIndex        = errors.Wrap(errBlobVerification, "incorrect blob index")
 )
 
 // BlobAlignsWithBlock verifies if the blob aligns with the block.
-func BlobAlignsWithBlock(blob *ethpb.DeprecatedBlobSidecar, block blocks.ROBlock) error {
+func BlobAlignsWithBlock(blob blocks.ROBlob, block blocks.ROBlock) error {
```
Comment (terencechain) [inline beacon-chain/sync/verify/blob.go:19]: Note there are some verifications that are missing here and I will address them in my PR

File: testing/util/deneb.go [hunk 8]
```diff
@@ -67,22 +68,48 @@ func GenerateTestDenebBlockWithSidecar(t *testing.T, parent [32]byte, slot primi
...
+		KzgProof:          commitment,
+	}
+	pb.CommitmentInclusionProof = fakeEmptyProof(t, pb)
+	r, err := blocks.NewROBlobWithRoot(pb, root)
+	require.NoError(t, err)
+	return r
+}
+
+func fakeEmptyProof(_ *testing.T, _ *ethpb.BlobSidecar) [][]byte {
+	r := make([][]byte, 17)
```
Comment (terencechain) [inline testing/util/deneb.go:102]: Maybe 17 should be config in fieldparams. Sorry I didn't define that previously

File: cmd/beacon-chain/storage/options.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/prysmaticlabs/prysm/v4/cmd"
+	"github.com/urfave/cli/v2"
+)
+
+var (
+	// StatePath defines a flag to start the beacon chain from a give genesis state file.
+	BlobStoragePath = &cli.PathFlag{
+		Name:  "blob-path",
+		Usage: "Location for blob storage. Default location will be a 'blobs' directory next to the beacon db.",
+	}
```
[the remark this answers is not in the document]
Comment (kasey) [reply cmd/beacon-chain/storage/options.go:14]: I'll look into that but it could be complicated given that it is based on another flag value - so there would need to be some variant of `Usage` that was lazy evaluated, so that it could access the set value of another flag.

File: beacon-chain/sync/verify/blob.go (hunk 2 above)
Comment (kasey) [reply beacon-chain/sync/verify/blob.go:23]: oops ty

File: beacon-chain/sync/blobs_test.go (hunk 4 above)
Comment (kasey) [reply beacon-chain/sync/blobs_test.go:120]: oh nice, I missed that we have that helper!

File: testing/util/deneb.go (hunk 8 above)
Comment (kasey) [reply testing/util/deneb.go:102]: My thinking was we'd generate real proofs as the pieces of this puzzle come together and throw this fakeEmptyProof helper away.

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 5 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/round_robin.go:169]: good catch, will add a `continue`.

Comment (kasey) [reply beacon-chain/sync/initial-sync/round_robin.go:169]: oh wait I forgot - if there is an error, `verified` will be nil, and the for loop won't do anything. But I can add a continue anyway since that could be confusing.

File: beacon-chain/db/filesystem/ephemeral.go (hunk 3 above)
Comment (kasey) [reply beacon-chain/db/filesystem/ephemeral.go:17]: In pure go a `_test.go` file can export symbols to other test packages, but I think we still haven't figured out a good way to do that with Bazel cc @prestonvanloon 

Cheap trick to enforce this: `NewEphemeralBlobStorage(_ *testing.T)`.

Comment (kasey) [reply beacon-chain/db/filesystem/ephemeral.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/977e5bafc7830fb9d6b30949db32cbe315bc8a33

File: beacon-chain/sync/blobs_test.go (hunk 4 above)
Comment (kasey) [reply beacon-chain/sync/blobs_test.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/98e5278f323794e2862cbbe54c40d8da4c0c1154

File: beacon-chain/db/filesystem/blob.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+// BlobStorage is the concrete implementation of the filesystem backend for saving and retrieving BlobSidecars.
+type BlobStorage struct {
+	fs afero.Fs
+}
+
+// Save saves blobs given a list of sidecars.
+func (bs *BlobStorage) Save(sidecar blocks.VerifiedROBlob) error {
+	fname := namerForSidecar(sidecar)
+	sszPath := fname.ssz()
```
Comment (terencechain) [inline beacon-chain/db/filesystem/blob.go:49]: `.ssz()` returns the path, but from the first look it made it seems like it's returning the ssz object itself. Dont feel strongly about this, but it may be more straightforward just to name it `.path()`, and partial for `.patialPath()`

File: beacon-chain/db/filesystem/blob.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		return err
+	}
+	if exists {
+		// TODO: should it be an error to save a blob that already exists?
+		return nil
+	}
+
+	// Serialize the ethpb.BlobSidecar to binary data using SSZ.
+	sidecarData, err := ssz.MarshalSSZ(sidecar)
```
Comment (terencechain) [inline beacon-chain/db/filesystem/blob.go:60]: Is this better than calling directly `sidecar. ssz.MarshalSSZ()`?

File: beacon-chain/db/filesystem/blob.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+		return v, err
+	}
+	s := &ethpb.BlobSidecar{}
+	if err := s.UnmarshalSSZ(encoded); err != nil {
+		return v, err
+	}
+	ro, err := blocks.NewROBlobWithRoot(s, root)
+	if err != nil {
+		return blocks.VerifiedROBlob{}, err
+	}
```
Comment (terencechain) [inline beacon-chain/db/filesystem/blob.go:112]: Another idea is to just save `blocks.VerifiedROBlob` in DB so we don't need to do this step. We will need to add `VerifiedROBlob` for fastssz. Not sure if it's worth it

File: beacon-chain/sync/subscriber_blob_sidecar.go [hunk 13]
```diff
@@ -4,29 +4,30 @@ import (
...
-	"github.com/prysmaticlabs/prysm/v4/beacon-chain/core/feed"
-	opfeed "github.com/prysmaticlabs/prysm/v4/beacon-chain/core/feed/operation"
-	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
 	"google.golang.org/protobuf/proto"
 )
 
 func (s *Service) blobSubscriber(ctx context.Context, msg proto.Message) error {
-	b, ok := msg.(*eth.SignedBlobSidecar)
+	b, ok := msg.(blocks.VerifiedROBlob)
```
Comment (terencechain) [inline beacon-chain/sync/subscriber_blob_sidecar.go:12]: I think this should be `ROBlob` instead of `VerifiedROBlob`. It's not verified yet right?

File: beacon-chain/db/filesystem/blob.go (hunk 12 above)
Comment (kasey) [reply beacon-chain/db/filesystem/blob.go:111]: We do save `VerifiedROBlob` in the db, but remember that from an ssz pov it's just a wrapper around `ethpb.BlobSidecar`. I did consider giving `ROBlob`/`VerifiedROBlob` `UnmarshalSSZ` methods, so they could directly unmarshal ssz values. The only reason they can't do it out of the box is that they need to initialize the internal `*ethpb.BlobSidecar` value to a non-nil instance, so the helper would just initialize the embedded value and then call unmarshal on it.

But I thought it was cleaner to preserve the rule that only the `verification` package gets to create `VerifiedROBlob` (in a non-error path) which we can work on enforcing with a static analyzer. So we will keep something like `verification.BlobSidecarNoop` that *only* the db is allowed to use.

File: beacon-chain/sync/subscriber_blob_sidecar.go (hunk 13 above)
Comment (kasey) [reply beacon-chain/sync/subscriber_blob_sidecar.go:12]: I thought the way `subscribeStaticWithSubnets` works is:
- first calls `validator` (`type wrappedVal func(context.Context, peer.ID, *pubsub.Message) (pubsub.ValidationResult, error)`), which would be sync service's `validateBlob` in this instance
- validator sets the libp2p `Message.Validator` value to the result of performing validation
- if validation is successful, `handle` (`type subHandler func(context.Context, proto.Message) error`) gets called, and it responsible for casting `Message.Validator` to the type that the `validator` sets.

So I thought it makes sense for the validation function to be responsible for "upgrading" ROBlob->VerifiedROBlob once it passes validation. Then `handle` only gets called with the verified blob.

Comment (terencechain) [reply beacon-chain/sync/subscriber_blob_sidecar.go:12]: My bad. I thought this was part of the validator handler. It's correct as it is now

File: beacon-chain/blockchain/process_block.go [hunk 14]
```diff
@@ -556,55 +572,23 @@ func (s *Service) isDataAvailable(ctx context.Context, root [32]byte, signed int
...
 	}
+
 	nc := s.blobNotifiers.forRoot(root)
 	for {
 		select {
 		case idx := <-nc:
-			found[idx] = struct{}{}
-			if len(found) != expected {
+			delete(missing, idx)
+			if len(missing) > 0 {
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/process_block.go:588]: Why do we do this? I dont understand how IsDataAvailable works. We wait for all of the missing data to be available then delete the blob on line 591?

File: cmd/beacon-chain/storage/options.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply cmd/beacon-chain/storage/options.go:16]: OK -- my personal opinion is that it would be better to have the full path in the help text but it doesn't seem very feasible. 

I think your implementation matches how others do it (ex: geth's keystore directory is the same) 

https://github.com/ethereum/go-ethereum/blob/5bf8769fb081c5f17f23b1b55e508789f0a8e632/node/config.go#L74

File: beacon-chain/blockchain/process_block.go [hunk 15]
```diff
@@ -529,11 +532,24 @@ func (s *Service) runLateBlockTasks() {
...
 }
 
+func missingIndices(bs *filesystem.BlobStorage, root [32]byte, expected int) (map[uint64]struct{}, error) {
+	indices, err := bs.Indices(root)
+	if err != nil {
+		return nil, err
+	}
+	missing := make(map[uint64]struct{}, expected)
+	for i := uint64(0); i < uint64(expected); i++ {
+		if !indices[i] {
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/blockchain/process_block.go:542]: indices is a fixed size, always equal to `MAX_BLOBS_PER_BLOCK`.
expected comes from blocks, and there are validation rules that enforce the kzg_commitments field to be `< MAX_BLOBS_PER_BLOCK`, eg in gossip validation we have `[REJECT] The sidecar's index is consistent with MAX_BLOBS_PER_BLOCK -- i.e. blob_sidecar.index < MAX_BLOBS_PER_BLOCK.`.

Usually we can assume ssz encoding acts as a validity check - ie you couldn't unmarshal a Vector beyond its fixed size, or a List exceeding its max size. But this field is weird, because the ssz is defined to match the EL theoretical max size (4096) not the current CL size limit (6). So maybe we should still do a defensive check here.

File: beacon-chain/db/filesystem/blob.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+)
+
+var (
+	errIndexOutOfBounds = errors.New("blob index in file name > MaxBlobsPerBlock")
+)
+
+const (
+	sszExt       = "ssz"
+	partExt      = "part"
+	blobLockPath = "blob.lock"
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/db/filesystem/blob.go:25]: yeah I removed some other code and missed that, ty

File: beacon-chain/p2p/gossip_topic_mappings.go [hunk 17]
```diff
@@ -21,7 +21,7 @@ var gossipTopicMappings = map[string]proto.Message{
 	SyncContributionAndProofSubnetTopicFormat: &ethpb.SignedContributionAndProof{},
 	SyncCommitteeSubnetTopicFormat:            &ethpb.SyncCommitteeMessage{},
 	BlsToExecutionChangeSubnetTopicFormat:     &ethpb.SignedBLSToExecutionChange{},
-	BlobSubnetTopicFormat:                     &ethpb.SignedBlobSidecar{},
+	BlobSubnetTopicFormat:                     &ethpb.BlobSidecar{},
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/p2p/gossip_topic_mappings.go:24]: There are no active devnets, and none going forward should use `SignedBlobSidecar`, and none of the testnets have forked.

File: beacon-chain/db/filesystem/blob.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+		}
+		parts := strings.Split(name, ".")
+		if len(parts) != 2 {
+			continue
+		}
+		u, err := strconv.ParseUint(parts[0], 10, 64)
+		if err != nil {
+			return mask, errors.Wrapf(err, "unexpected directory entry breaks listing, %s", parts[0])
+		}
+		if u > fieldparams.MaxBlobsPerBlock {
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/db/filesystem/blob.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/680636ffe3751862b1ac7e8a8876cc381e615199

File: beacon-chain/blockchain/process_block.go (hunk 14 above)
Comment (kasey) [reply beacon-chain/blockchain/process_block.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/f746a3af31c9bc40c5e349ef8d0932814ccf1cfe

File: beacon-chain/blockchain/process_block.go [hunk 19]
```diff
@@ -529,11 +532,24 @@ func (s *Service) runLateBlockTasks() {
 	}
 }
 
+func missingIndices(bs *filesystem.BlobStorage, root [32]byte, expected int) (map[uint64]struct{}, error) {
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/blockchain/process_block.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/f3b3afcb0f1a8d7400a2dd8b043285ae1151cc32

File: beacon-chain/blockchain/process_block.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/blockchain/process_block.go]: defensive check here:
https://github.com/prysmaticlabs/prysm/pull/13190/commits/f3b3afcb0f1a8d7400a2dd8b043285ae1151cc32
and I think passing in the expected commitment slice instead of the length of it might make things slightly more clear:
https://github.com/prysmaticlabs/prysm/pull/13190/commits/71e65af6839d5726bd023fab5d3b33898e385b92

File: beacon-chain/blockchain/process_block.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/blockchain/process_block.go]: comment: https://github.com/prysmaticlabs/prysm/commit/f3b3afcb0f1a8d7400a2dd8b043285ae1151cc32

File: beacon-chain/db/filesystem/blob.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/db/filesystem/blob.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/8cde358af3fbc44780041a15a4fb78c0edfb1523

File: beacon-chain/db/filesystem/blob.go [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+// Save saves blobs given a list of sidecars.
+func (bs *BlobStorage) Save(sidecar blocks.VerifiedROBlob) error {
+	fname := namerForSidecar(sidecar)
+	sszPath := fname.ssz()
+	exists, err := afero.Exists(bs.fs, sszPath)
+	if err != nil {
+		return err
+	}
+	if exists {
+		// TODO: should it be an error to save a blob that already exists?
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/db/filesystem/blob.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/0742ebc03afbeee881906b4d278842f515871f4a

File: testing/util/deneb.go (hunk 8 above)
Comment (kasey) [reply testing/util/deneb.go:101]: I was going to say it feels goofy adding this since it's more an empirical fact than a config value, but it turns out the spec does treat it as a [Preset](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md#preset), so I'll go ahead and add it.

Comment (kasey) [reply testing/util/deneb.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/ccf828f06ec2aced2d4695d3cd2806e60391e5e9

File: beacon-chain/sync/initial-sync/round_robin.go (hunk 5 above)
Comment (kasey) [reply beacon-chain/sync/initial-sync/round_robin.go:169]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/d25b7d27edf28cb4e639c33e64a7f766c5b7904e

File: beacon-chain/db/filesystem/blob.go (hunk 10 above)
Comment (kasey) [reply beacon-chain/db/filesystem/blob.go]: https://github.com/prysmaticlabs/prysm/pull/13190/commits/6239576594a5bf3876cf00ac39da8a96a439ee8d

File: beacon-chain/sync/validate_blob.go [hunk 21]
```diff
@@ -159,35 +164,39 @@ func (s *Service) validateBlobPostSeenParent(ctx context.Context, sBlob *eth.Sig
 	if err != nil {
 		return pubsub.ValidationIgnore, err
 	}
-	if err := verifyBlobSignature(parentState, sBlob); err != nil {
-		log.WithError(err).WithFields(blobFields(blob)).Debug("Failed to verify blob signature")
-		return pubsub.ValidationReject, err
-	}
+	// TODO: replace this with verification routine
```
Comment (potuz) [inline beacon-chain/sync/validate_blob.go:167]: Why not call here the proof verification helper?

File: beacon-chain/sync/validate_blob.go [hunk 22]
```diff
@@ -159,35 +164,39 @@ func (s *Service) validateBlobPostSeenParent(ctx context.Context, sBlob *eth.Sig
...
+	/*
+		if err := verifyBlobSignature(parentState, blob); err != nil {
+			log.WithError(err).WithFields(blobFields(blob)).Debug("Failed to verify blob signature")
+			return pubsub.ValidationReject, err
+		}
+	*/
 
 	// [IGNORE] The sidecar is the only sidecar with valid signature received for the tuple (sidecar.block_root, sidecar.index).
```
Comment (potuz) [inline beacon-chain/sync/validate_blob.go:175]: This comment probably needs to change

File: beacon-chain/sync/validate_blob.go [hunk 23]
```diff
@@ -159,35 +164,39 @@ func (s *Service) validateBlobPostSeenParent(ctx context.Context, sBlob *eth.Sig
...
-	if s.hasSeenBlobIndex(blob.BlockRoot, blob.Index) {
+	if s.hasSeenBlobIndex(blob.BlockRootSlice(), blob.Index) {
 		return pubsub.ValidationIgnore, nil
 	}
 
 	// [REJECT] The sidecar is proposed by the expected proposer_index for the block's slot in the context of the current shuffling (defined by block_parent_root/slot)
-	parentState, err = transition.ProcessSlotsUsingNextSlotCache(ctx, parentState, parentRoot[:], blob.Slot)
+	parentState, err = transition.ProcessSlotsUsingNextSlotCache(ctx, parentState, parentRoot[:], blob.Slot())
```
Comment (potuz) [inline beacon-chain/sync/validate_blob.go:181]: Why are we grabbing a state to do this, can't we use the proposer index cache instead?

Comment (kasey) [reply beacon-chain/sync/validate_blob.go:181]: Agreed, but this line only changed because I updated the slot accessor - let's resolve that in the upcoming verification PR: https://github.com/prysmaticlabs/prysm/pull/13197

File: beacon-chain/sync/validate_blob.go (hunk 21 above)
Comment (kasey) [reply beacon-chain/sync/validate_blob.go:167]: We'll do that in https://github.com/prysmaticlabs/prysm/pull/13197

File: beacon-chain/sync/validate_blob.go (hunk 23 above)
Comment (terencechain) [reply beacon-chain/sync/validate_blob.go:181]: I can handle this check in my gossip PR as well

**Review by terencechain — APPROVED** [review]

File: beacon-chain/blockchain/process_block.go [hunk 24]
```diff
@@ -529,11 +533,43 @@ func (s *Service) runLateBlockTasks() {
 	}
 }
 
+// missingIndices uses the expected commitments from the block to determine
+// which BlobSidecar indices would need to be in the database for DA success.
+// It returns a map where each key represents a missing BlobSidecar index.
+// An empty map means we have all indices; a non-empty map can be used to compare incoming
+// BlobSidecars against the set of known missing sidecars.
+func missingIndices(bs *filesystem.BlobStorage, root [32]byte, expected [][]byte) (map[uint64]struct{}, error) {
```
Comment (potuz) [inline beacon-chain/blockchain/process_block.go:541]: Is there a test for this one?

File: consensus-types/blocks/roblob.go [hunk 25]
```diff
@@ -62,3 +62,31 @@ func (b *ROBlob) BodyRoot() [32]byte {
 func (b *ROBlob) ProposerIndex() primitives.ValidatorIndex {
 	return b.SignedBlockHeader.Header.ProposerIndex
 }
+
+// BlockRootSlice returns the block root as a byte slice. This is often more conveninent/concise
+// than setting a tmp var to BlockRoot(), just so that it can be sliced.
+func (b *ROBlob) BlockRootSlice() []byte {
```
Comment (potuz) [inline consensus-types/blocks/roblob.go:68]: I am a little concerned with the possible usages of this function since there may be a data race when the root is being set.

File: beacon-chain/db/filesystem/blob.go [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+// Save saves blobs given a list of sidecars.
+func (bs *BlobStorage) Save(sidecar blocks.VerifiedROBlob) error {
+	fname := namerForSidecar(sidecar)
+	sszPath := fname.path()
+	exists, err := afero.Exists(bs.fs, sszPath)
+	if err != nil {
+		return err
+	}
+	if exists {
+		log.WithFields(logging.BlobFields(sidecar.ROBlob)).Debug("ignorning a duplicate blob sidecar Save attempt")
```
Comment (prestonvanloon) [inline beacon-chain/db/filesystem/blob.go:56]:
```suggestion
		log.WithFields(logging.BlobFields(sidecar.ROBlob)).Debug("ignoring a duplicate blob sidecar Save attempt")
```

**Review by prestonvanloon — DISMISSED** [review]
LGTM - found a typo though

File: beacon-chain/sync/rpc_beacon_blocks_by_root.go [hunk 27]
```diff
@@ -180,39 +181,35 @@ func (s *Service) sendAndSaveBlobSidecars(ctx context.Context, request types.Blo
...
-	for _, blob := range blobs {
-		indices[blob.Index] = struct{}{}
-	}
-	return indices
+	return requestsForMissingIndices(have, count, blockRoot), nil
 }
 
 // Helper function to filter out known indices.
-func filterUnknownIndices(knownIndices map[uint64]struct{}, count int, blockRoot [32]byte) []*eth.BlobIdentifier {
+func requestsForMissingIndices(haveIndices [fieldparams.MaxBlobsPerBlock]bool, count int, blockRoot [32]byte) []*eth.BlobIdentifier {
```
Comment (saolyn) [inline beacon-chain/sync/rpc_beacon_blocks_by_root.go:207]: `haveIndices` isn't great, could this not just be called `indices`? It makes the variable `has` on line 198 is a bit unclear

File: beacon-chain/sync/validate_blob.go [hunk 28]
```diff
@@ -133,22 +134,26 @@ func (s *Service) validateBlob(ctx context.Context, pid peer.ID, msg *pubsub.Mes
 
 	blobSidecarArrivalGossipSummary.Observe(float64(sinceSlotStartTime.Milliseconds()))
 
-	msg.ValidatorData = sBlob
+	vblob, err := verification.BlobSidecarNoop(blob)
+	if err != nil {
+		return pubsub.ValidationReject, errors.New("TODO: undefined verification error")
```
Comment (saolyn) [inline beacon-chain/sync/validate_blob.go:139]: Is this going to be handled in another PR?

File: consensus-types/blocks/roblob.go (hunk 25 above)
Comment (kasey) [reply consensus-types/blocks/roblob.go:68]: The root is *only* set in the constructor.

File: beacon-chain/sync/rpc_beacon_blocks_by_root.go (hunk 27 above)
Comment (kasey) [reply beacon-chain/sync/rpc_beacon_blocks_by_root.go]: What do you think of this? https://github.com/prysmaticlabs/prysm/pull/13190/commits/f4af7dc5b5dee43abe878db48f0eabd369312ea6

File: beacon-chain/sync/validate_blob.go (hunk 28 above)
Comment (kasey) [reply beacon-chain/sync/validate_blob.go:139]: yes, we'll get rid of the "fake" verification code in the next PR. Currently this is unreachable as `BlobSidecarNoop` never returns an error.

File: beacon-chain/blockchain/process_block.go (hunk 24 above)
Comment (kasey) [reply beacon-chain/blockchain/process_block.go:541]: added https://github.com/prysmaticlabs/prysm/pull/13190/commits/121ef9052eb983088e09ec2efeeae7ee322d9ad0

**Review by saolyn — DISMISSED** [review]
LGTM

**Review by terencechain — APPROVED** [review]
